### PR TITLE
FFI plugins

### PR DESCRIPTION
**Pull request by dcharkes**
Add support for `flutter create --template=plugin_ffi` and `flutter run` support for `ffiPlugin: true` in `pubspec.yaml`.

This addresses 2 issues:

1. Currently, the way to do FFI with Flutter is to use `flutter create --template=plugin`. However, this generates boilerplate for plugin registration and example code for method channels which go unused.
2. We've see dozens of issues with building, linking, and bundling native code. The FFI plugin template should be the sample that works for all platforms for users to reference or use as a starting point.

[Design doc].
[NDK versioning doc].

Some relevant issues:

* https://github.com/flutter/flutter/issues/78819
* https://github.com/flutter/flutter/issues/77790

Included in this PR:

* Building shared C source code as part of the native build for platforms Android, iOS, Linux desktop, MacOS desktop, and Windows desktop.
* Sample code doing a synchronous FFI call.
* Sample code doing a long running synchronous FFI call on a helper isolate.
* Use of `package:ffigen` to generate the bindings.

Not included in this PR:

* Bundling closed source libraries. (MacOS, iOS, and Android require copying to the right place. Linux and Windows require referencing from CMake files.)
* Building closed source libraries with the right compiler options (most notably rpath and include path).
* Support for `flutter test`. This requires a `bin/setup.dart` to invoke a native build outside `flutter run` or `flutter build`.
* Support for using the generated package as Dart standalone. This requires a `bin/setup.dart` to invoke a native build.
* Linking against `dart_api_dl.h` to be able to invoke functions from `dart_api.h`.
* Sample code doing an asynchronous callback through native ports form `dart_api.h`.
* Compatibility between the `plugin` and `plugin_ffi` templates. The example apps (`example/lib/main.dart`) are different.

Test this PR locally without building Flutter by running the Flutter commands from source (replace platforms for your host OS). 

```sh
$ dart <PATH_TO_FLUTTER_CHECKOUT>/packages/flutter_tools/bin/flutter_tools.dart create -t plugin_ffi --platforms=macos,ios,android my_ffi_plugin
$ cd my_ffi_plugin/example/
$ dart <PATH_TO_FLUTTER_CHECKOUT>/packages/flutter_tools/bin/flutter_tools.dart run -d macos
```

For context, previous PR (reverted):

* https://github.com/flutter/flutter/pull/94101

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Design doc]: https://docs.google.com/document/d/1vMw-W6L-r-uL1SUgzDh6J5rmtqum7SZdO9khWACrZJI/edit?usp=sharing&resourcekey=0-vvjEK88N-pApyKTPWJ2IZQ
[NDK versioning doc]: https://docs.google.com/document/d/1sljfqHpI5m4uxsL7B89yhYluO97Mz1YKyA0VgzzXyZ4/edit?usp=sharing&resourcekey=0-Ts9Rk8jbJor7WBMMiehKWg
[Internal design doc]: http://go/flutter-dart-ffi-native-libraries
[Android integration test]: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart

### CODE REVIEW

File: packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl
Comment (blasten) [inline packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl:42]: I left a comment in the doc related to this part.  

The problem is that in Flutter the Gradle/AGP versions to be used are actually specified in the app instead of the plugin. In fact, the AGP/Gradle version doesn't have any effect unless the plugin is built as aar with `flutter build aar`.

**Comment by leisim** [comment]
Will support for prebuilt libraries be added in the future? It's not only important for closed-source libraries but also if the compiler is too difficult to setup on the users machine (for example for Rust libraries)

**Comment by dcharkes** [comment]
> Will support for prebuilt libraries be added in the future? It's not only important for closed-source libraries but also if the compiler is too difficult to setup on the users machine (for example for Rust libraries)

It is the intention to support that use case as well. At this point we're not entirely sure whether that warrants a new template, or whether a sample and documentation is enough. (Some pointers based on the source template: add the precompiled dl path to Windows Linux `CMakeLists.txt` in `_bundled_libraries`, copy the dl to MacOS iOS Frameworks folder, copy the dl to jniLibs.)

Also, it would require package authors to upload the binaries for all (18) target platforms (hardware * OS combination) to pub. We're exploring if there are any obstacles for doing that.

**Comment by leisim** [comment]
Building the binaries for all platforms is no problem but. More documentation would be very helpful. Especially how to:
- put shared libraries in the correct place to be found at runtime
- add multiple binaries for the same os (ios_arm64, ios_arm64_sim etc.)
- make sure the binaries don't get stripped

**Review by blasten — APPROVED** [review]

File: packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl
Comment (dcharkes) [reply packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl:42]: I have been able to verify this behavior. The `android.ndkVersion` in the FFI plugin `build.gradle` is ignored when building a Flutter app that imports it. The `android.ndkVersion` in the apps `app/build.gradle` is used.

We'll need to find a solution for this behavior.

File: dev/devicelab/lib/tasks/plugin_tests.dart
Comment (christopherfujino) [inline dev/devicelab/lib/tasks/plugin_tests.dart:49]: unrelated to this test, what is the user experience if I run `flutter test` on an app with an FFI plugin? Does it just fail to lookup the symbol?

File: packages/flutter_tools/lib/src/commands/create.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/create.dart:166]: this could still be within the `else {`, right? since if we've set `template` on line 165 then the value of `detectedProjectType` won't be used anyway, right?

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [inline packages/flutter_tools/gradle/flutter.gradle:466]: nice

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart:129]: Nit: Remove the ", if any"

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart:130]: Does this actually need to be nullable? Not present should be equivalent to `false`.

(Same in the other classes below.)

File: packages/flutter_tools/lib/src/plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/plugins.dart:58]: Nit: "implemented with"

File: packages/flutter_tools/templates/plugin_ffi/README.md.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/README.md.tmpl:16]: Nit: missing period.

File: packages/flutter_tools/templates/plugin_ffi/README.md.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/README.md.tmpl:19]: Same

File: packages/flutter_tools/templates/plugin_ffi/README.md.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/README.md.tmpl:26]: nit: s/./:/

File: packages/flutter_tools/templates/plugin_ffi/README.md.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/README.md.tmpl:38]: A likely more common case, and thus worth mentioning before this, is `ffiPlugin`+`dartPluginClass`.

E.g.,:

> This can be combined with `dartPluginClass`, such as when FFI is used for the
> implementation of one platform in a federated plugin:
> 
> ` ``` `yaml
>  plugin:
>    implements: some_other_plugin
>    platforms:
>      some_platform:
>        dartPluginClass: SomeClass
>        ffiPlugin: true
> ` ``` `

File: packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl:28]: s/users/clients/ in all of these comments.

File: packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl:1]: Please add comments like those I added to the other CMake files about ideally not changing this version.

File: packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl:11]: Apologies if we've discussed this and I forgot: why is this done differently from method channel plugins (where this is done in the generated file)? That would simplify both that file (since it would make the two paths the same) and this one. And it would provide one less way for people to break themselves when editing these files.

File: packages/flutter_tools/templates/plugin_ffi/macos.tmpl/Classes/projectName.c.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/macos.tmpl/Classes/projectName.c.tmpl:1]: It's probably worth explicitly pointing to the podfile to explain why (e.g., "See the comment in [...].podfile for more information). Otherwise people may be confused if/when they want to add more sources, since they are much more likely to look at this file than the podfile.

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:14]: It would be nice to format this CMake file more like the Linux and Windows ones for consistency.

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:1]: Per previous-cycle review comments, I think "we" is confusing in a template. I would say "This is" rather than "we use"

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:6]: "clients" (or "customers")

Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:6]: It's actually worse than that; at least on Windows they would need to either:
- download a newer version of Visual Studio, or
- do surgery on the internals of their VS installation.

I would phrase this more strongly, as in the Windows and Linux plugin comments I added:
"Changing the CMake version will cause the plugin to fail to compile for some
customers of the plugin."

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/projectName.h.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/src.tmpl/projectName.h.tmpl:13]: Why "MYLIB"? It sounds like a placeholder. If it's supposed to be the name of their library we should populate it that way; if not it seems like it should be something that sounds generic rather than like a placeholder (e.g., "FFI_PLUGIN_EXPORT")

File: packages/flutter_tools/templates/plugin_ffi/windows.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_ffi/windows.tmpl/CMakeLists.txt.tmpl:1]: Same comments in this file.

File: packages/flutter_tools/templates/plugin_shared/pubspec.yaml.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_shared/pubspec.yaml.tmpl:12]: I'm not sure about this inconsistency; this will mean that (forever, not just right now) by default someone on master creating a platform channel plugin will work for ~everyone, but someone making an FFI plugin will get a plugin that works for almost nobody. I think we should consider making this hard-coded, or maybe making this "minFrameworkVersion" which is populated in code as the earlier of 2.11.0 and the framework version (if we want to exactly handle the version when it lands), with a TODO to switch it to hard-coded later.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Lots of small comments, but overall this is looking really good!

File: dev/devicelab/lib/tasks/plugin_tests.dart
Comment (dcharkes) [reply dev/devicelab/lib/tasks/plugin_tests.dart:49]: The FFI plugin template itself doesn't include any unit tests:

> `Test directory "test" not found.`

When adding a test, opening the dynamic library fails:

> `Shell: Invalid argument(s): Failed to load dynamic library '<...>':`

I created a tracking issue:

* https://github.com/flutter/flutter/issues/97056

File: packages/flutter_tools/lib/src/commands/create.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/commands/create.dart:166]: If it is in the else, the following code is dead code:

```dart
    if (detectedProjectType != null && template != detectedProjectType && metadataExists) {
      // We can only be definitive that this is the wrong type if the .metadata file
      // exists and contains a type that doesn't match.
      throwToolExit("The requested template type '${flutterProjectTypeToString(template)}' doesn't match the "
          "existing template type of '${flutterProjectTypeToString(detectedProjectType)}'.");
    }
```

`template != detectedProjectType` will never be true.

That means we do not detect inconsistent project types. For example having method channel `plugin` in the .metadata file and `plugin_ffi` as an argument. Not detecting inconsistent project types results in a garbled project state with files from two different templates mixed together.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (dcharkes) [reply packages/flutter_tools/lib/src/platform_plugins.dart:130]: Done.

Please note that `toMap()` will now not get an entry if `ffiPlugin: false`.

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl
Comment (dcharkes) [reply packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:6]: I've replaced the message with the same one you added on the other CMakeLists files:

```
# The Flutter tooling requires that developers have CMake 3.10 or later
# installed. You should not increase this version, as doing so will cause
# the plugin to fail to compile for some customers of the plugin.
```

File: packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl
Comment (dcharkes) [reply packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl:11]: We'd like to share the the CMakeLists.txt in src across all platforms.

We could standardise on always expecting a `CMakeLists.txt` file in `src` and standardise on a target name. However, if we have users with existing CMake builds that do not have their build file in precisely that location they cannot modify generated file that's generated every build.

Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin_ffi/linux.tmpl/CMakeLists.txt.tmpl:18]: I see, it wasn't clear to me that this was to allow alternate build structures. Maybe it's worth putting a comment here and on line 11 that they could be pointed to a different build?

**Comment by dcharkes** [comment]
> Lots of small comments, but overall this is looking really good!

@stuartmorgan I've addressed all of them. PTAL.

@christopherfujino I've replied to your comment above. Please let me know if you have any further comments.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (stuartmorgan) [inline packages/flutter_tools/gradle/flutter.gradle:424]: Typo: semantic

File: packages/flutter_tools/lib/src/commands/create_base.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/create_base.dart:404]: This is unused now, isn't it?

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart:101]: You can remove the `== true` from these now that you've made them non-nullable.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM! Thanks for all the work here :)

**Comment by dcharkes** [comment]
Thanks Stuart!

> Thanks for all the work here :)

Let's make our users happy!

@christopherfujino waiting on your approval (or delegation) as `flutter_tools` owner.

**Review by christopherfujino — APPROVED** [review]
LGTM! Thanks for landing this!

**Comment by GregoryConrad** [comment]
I have actually just ran into the issue @leisim prophesied; I am trying to build a Flutter library wrapping around a Rust library. Seeing as I cannot expect users of the library to have `rustup` installed, I plan on distributing precompiled versions of the library for all the targets.

For macOS/iOS, my best attempt was to create an XCFramework from the compiled rust static libraries (quite a few `.a` files!). After `lipo`-ing a the simulator and macOS static libs together, I was able to create the full XCFramework. For future reference, perhaps if someone is making some Rust/Flutter library documentation, I have this process thrown together in a [build script](https://github.com/GregoryConrad/mimir/blob/a97bd600fae1df545778324925b689cd79165395/packages/mimir/native/build-apple.sh). This takes care of all Apple platforms in one framework, which is very handy. However, when I go to include this Framework in my plugin, I [run into linker errors as seen in this SO post](https://stackoverflow.com/questions/74467824/creating-flutter-rust-ffi-plugin-with-an-xcframework-for-macos-ios).

@dcharkes mentions:
> (Some pointers based on the source template: add the precompiled dl path to Windows Linux CMakeLists.txt in _bundled_libraries, copy the dl to MacOS iOS Frameworks folder, copy the dl to jniLibs.)

I'm most interested in the macOS/iOS part here. Could you please provide a brief working example of how to go about this for macOS/iOS? I have tried for hours and cannot get the XCFramework & podspec file to work properly.

**Comment by dcharkes** [comment]
### 100 MB package version limit

@jonasfj We have our first user hit the 100 MB limit.

### Precompiled static libraries on MacOS/iOS in a xcframework

> I'm most interested in the macOS/iOS part here (for now). Could you please provide a brief working example of how to go about this for macOS/iOS? I have tried for hours and cannot get the XCFramework & podspec file to work properly.

I've got something slightly working for precompiled static libraries in an xcframework:

mylib_staticlib/macos/mylib_staticlib.podspec

```podspec
#
# To learn more about a Podspec see http://guides.cocoapods.org/syntax/podspec.html.
# Run `pod lib lint mylib_staticlib.podspec` to validate before publishing.
#
Pod::Spec.new do |s|
  s.name             = 'mylib_staticlib'
  s.version          = '0.0.1'
  s.summary          = 'A new flutter plugin project.'
  s.description      = <<-DESC
A new flutter plugin project.
                       DESC
  s.homepage         = 'http://example.com'
  s.license          = { :file => '../LICENSE' }
  s.author           = { 'Your Company' => 'email@example.com' }
  s.source           = { :path => '.' }
  s.source_files     = 'Classes/**/*'
  s.dependency 'FlutterMacOS'

  s.platform = :osx, '10.11'
  s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES' }
  s.vendored_libraries = 'Frameworks/libmylib_staticlib.a'
  s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/mylib_staticlib.xcframework/macos-arm64_x86_64/libmylib_staticlib.a" }
  s.swift_version = '5.0'
end
```

mylib_staticlib/macos/Classes/dummy_file.c

```c
// A dummy file so that the mylib_staticlib.framework is created.
```

Path of the framework with static libraries: mylib_staticlib/ios/Frameworks/mylib_staticlib.xcframework/ios-arm64_x86_64-simulator/libmylib_staticlib.a

One issue with doing it like this is that the ios_arm64 is compiled either against the iossimulator or against ios, and I wasn't able to figure out how to make that both work.

**Comment by leisim** [comment]
Check out the [Isar podspec](https://github.com/isar/isar/blob/main/packages/isar_flutter_libs/ios/isar_flutter_libs.podspec) and the [build script](https://github.com/isar/isar/blob/main/tool/build_ios.sh) for Rust.

It works really well even without -force_load. That being said it is still a huge pain to ship native libraries for all platforms and I really hope there will be a nice solution by Flutter. Ideally allowing static compilation into the Flutter native lib.

**Comment by dcharkes** [comment]
> Check out the [Isar podspec](https://github.com/isar/isar/blob/main/packages/isar_flutter_libs/ios/isar_flutter_libs.podspec) and the [build script](https://github.com/isar/isar/blob/main/tool/build_ios.sh) for Rust.
> 
> It works really well even without -force_load and supports all cpu/simulator configurations. That being said it is still a huge pain to ship native libraries for all platforms and I really hope there will be a nice solution by Flutter. Ideally allowing static compilation into the Flutter native lib.

As the end goal, we would like static compilation into the compiled Dart code, that means we can use the native linker to get rid off all native code not referenced. Tracked in:

* https://github.com/dart-lang/sdk/issues/49418

**Comment by GregoryConrad** [comment]
@leisim Thanks for that! I didn't think of using a regular plugin package instead of an FFI plugin to actually integrate everything but it looks like it works smoothly. Also, thank you both for the speedy response!

### Static Linking w/ Dart
This would be absolutely fantastic, as the generated static libraries I am dealing with are quite large (~100-200 mb _per platform & architecture_) and I am sure they have plenty of dead code not needed by my Dart library in them (caused by a large number of dependencies in the Rust library I am wrapping).

### 100 MB Limit
As I mentioned in the previous paragraph, the static libraries I am dealing with are exceedingly large, and the entire upload could easily be above 1 GB. If you do consider changing the pub limit, perhaps it would be worth considering having a separate upload for static/native libraries and then link to them in the main upload (or perhaps a different way to handle them altogether)? Maybe, if Dart does get static linking, the theoretical separate upload could just be for native libraries that are statically linked automatically by Flutter/Dart? That could be a really elegant solution.

**Comment by leisim** [comment]
@GregoryConrad 100-200MB sounds incredibly large. Did you check the [steps to reduce Rust binary size](https://github.com/johnthagen/min-sized-rust)? It's quite expensive to load such a fast binary at runtime.

**Comment by GregoryConrad** [comment]
In case anyone else stumbles into this, I got FFI working with my cross-platform XCFramework using `ffiPlugin: true` (thanks to @leisim's great example with Isar!). I can't believe I didn't realize this, but my XCFramework was getting tree-shaken. Here's how I fixed it (same exact code for macOS and iOS, but _cannot_ be symbolically linked unfortunately):

- `my_package.podspec`
```ruby
Pod::Spec.new do |spec|
  spec.name          = 'my_package'
  spec.version       = '0.0.1'
  spec.license       = { :file => '../LICENSE' }
  spec.homepage      = ''
  spec.authors       = { 'Your Name' => 'blabla@example.com' }
  spec.summary       = 'Some summary'

  spec.source              = { :path => '.' }
  spec.source_files        = 'Classes/**/*'
  spec.public_header_files = 'Classes/**/*.h'
  spec.vendored_frameworks = 'Frameworks/MyPackage.xcframework'

  spec.ios.deployment_target = '11.0'
  spec.osx.deployment_target = '10.11'
end
```
- `Classes/binding.h`
```
void enforce_binding();
```
- `Classes/EnforceBinding.swift`
```swift
public func dummyMethodToEnforceBundling() {
    enforce_binding() // disable tree shaking
}
```
- `Frameworks/MyPackage.xcframework` needs to have a `void enforce_binding() {}` function. I did this in Rust with:
```rust
/// Enforce the binding for this library (to prevent tree-shaking)
#[no_mangle]
pub extern "C" fn enforce_binding() {}
```

**Comment by jonasfj** [comment]
I'm not sure we should tweak the 100MB limit.

If you're things are really this big, then maybe:
 * You should download platform specific binary blobs in the build step (@dcharkes this certainly makes sense if binary blobs are 200mb per platform).
 * Work to make your package smaller,
 * Split the package into multiple packages, if this makes sense, so that users don't have to download all at once.

**Pull request merged**
